File: quiz/deno/src/statemaschine/states/quiz/participate/run-quiz-state.ts
import { Quiz } from '@src/quiz/quiz.ts';
import { IState } from '@states/state.ts';
import { InitState } from '@states/init-state.ts';
import { keypress } from '@cliffy/keypress/mod.ts';

export class RunQuizState implements IState {
	constructor(private readonly quiz: Quiz) {}

	async run() {
		const questions = this.quiz.questions;

		const answers = new Map<string, boolean>();

		for (const question of questions) {
			const answer = await question.ask();
			if (question.checkAnswer(answer)) {
				console.log(
					'%cCorrect!',
					'color: #0f0; font-weight: bold;',
				);
				answers.set(question.id, true);
			} else {
				console.log(
					'%cWrong!',
					'color: #f00; font-weight: bold;',
				);
				answers.set(question.id, false);
			}
			console.log('Press any key to continue...');
			await keypress();
		}

		this.quiz.addResponse(answers);
		await this.quiz.save();

		return new InitState();
	}
}

File: quiz/go/statemaschine/states/run-quiz-state.go
package states

import (
	"github.com/DoktorLenz/kp-wise23-24/quiz/go/quiz"
	"github.com/DoktorLenz/kp-wise23-24/quiz/go/utils"
)

type RunQuizState struct {
	quiz *quiz.Quiz
}

func (state *RunQuizState) Run() IState {

	questions := state.quiz.Questions
	answers := make(map[string]bool)

	for _, question := range questions {
		answer, err := question.Ask()
		if err != nil {
			panic(err)
		}
		if answer {
			utils.Prompt("Correct!")
			answers[question.GetID()] = true
		} else {
			utils.Prompt("Wrong!")
			answers[question.GetID()] = false
		}
		utils.Prompt("")
		utils.AwaitEnter()
	}

	state.quiz.AddResponse(answers)
	state.quiz.Save()

	utils.Clear()
	utils.Prompt("You have finished the quiz!")
	utils.Prompt("Thank you for participating!")
	utils.Prompt("")
	utils.AwaitEnter()

	return &InitState{}
}
